### PR TITLE
generative functor: the parameter name is also optional

### PR DESCRIPTION
**Pull request by trefis**
This feels like a better representation.

I also think that:
- the bug in #8891 could not have been written with this representation.
- the code in translmod that creates path like `F(*).Foo` (probably) wouldn't have been written in the first place?
- ocamldoc's implementation (which I haven't changed) would have handled generative functors specially, instead of just using `*` as the parameter name

These last two being speculative of course.

### CODE REVIEW

File: typing/typemod.ml
Comment (trefis) [inline typing/typemod.ml]: @garrigue (or @lpw25): I don't think the signature of an generative functor needs to be rechecked on applications. Can you confirm?

**Comment by gasche** [comment]
I agree with the approach and I think it goes in the good direction, but is it going far enough? We still use strings to represent argument names, with `"_"` used to represent anonymous parameters, and this is kind of wrong. If we are going to change the Parsetree and break all parsetree consumers again, maybe we could be slightly more perfectionist about it?

Some remarks:

1. I'm not fond of the use of an option to distinguish generative from applicative arguments. I think that `()` semantically means something else than "there is no argument", which is what `None` suggest. Maybe we could have a proper datatype?

2. I don't think that `"_"` is a good representation of the syntax that the user writes as `_` (as in `functor (_ : S) ...`); but there I think that an option would do fine.

3. I'm not sure of whether we should distinguish between `functor (_ : S) -> T` (a dependent anonymous abstraction) and `S -> T` (a non-dependent abstraction). Currently they are represented in the same way, except that the `"_"` has a ghost location in the second case. I think that it would make sense to represent them differently in the parsetree, but probably not in the typedtree.

One possible representation would be something like

```ocaml
type functor_arg =
  | Generative
  | Anonymous of module_type
  | Named of string loc * module_type
```

**Comment by trefis** [comment]
The remarks seem reasonable, your suggested `functor_arg` type however is a bit disappointing (it's the functor that is generative, not its argument).

However, if we're proceeding with these bold changes, then I'm also going to add #6662 to the mix.

**Comment by yallop** [comment]
> However, if we're proceeding with these bold changes, then I'm also going to add #6662 to the mix.

Indeed, this seems like a good time to resolve #6662, since the main concern about that proposal was the (compatibility-breaking) change to the parse tree, which is going to change here in any case.

(On the other hand, if you also implement @gasche's [suggestion](https://github.com/ocaml/ocaml/issues/6662#issuecomment-473038355) in the #6662 discussion of adding `and`-binding for modules, which I also favour, it'd be better to do so in a separate PR.)

**Comment by trefis** [comment]
> (On the other hand, if you also implement @gasche's suggestion in the #6662 discussion of adding and-binding for modules, which I also favour, it'd be better to do so in a separate PR.)

Agreed. That one seems useful but out of scope. `module _ =`, `module () = ..` and `(... : ())` (as well as adding `module Ignore : sig end -> ()` to `Stdlib`?) seem very much in scope.

**Comment by lpw25** [comment]
I doubt it's particularly helpful but I have a branch implementing the `module _ =` part of #6662 here:

https://github.com/lpw25/ocaml/tree/underscore_module

I think that including that in this PR makes sense.

I'm less sure about including the `module () = ...` stuff. Even though I suggested it I'm not 100% sure the design is quite right, and I don't think this PR should get tied up in those discussions.

**Comment by lpw25** [comment]
> I think that including that in this PR makes sense.

By which I mean including that feature not including my code. My implementation is so out-of-date its probably quicker to start from scratch -- also my branch does "clever" things with recursive module bindings that in retrospect are a mistake.

**Comment by gasche** [comment]
> it's the functor that is generative, not its argument

No strong opinion here but I see nothing wrong with calling `()` a "generative argument". I am quite sure it is correct to call `functor () ->` a "generative (functor) abstraction", so "generative argument" makes sense to denote the `()` in that position; and then `Generative_arg` or `Generative` make sense (to me) as constructor names.

To make a comparison, we call `(lazy p)` a "lazy pattern", although it is in fact not adding lazyness, it is forcing -- it is observing/destructing laziness. The `()` argument of a generative functor is observing/witnessing generativity.

**Comment by trefis** [comment]
Alright, I just pushed a fairly big commit.

On the module binding side of things, I wrapped the name in an option, everything was fairly straightforward (although tedious).

For the functors parameter representation change, I replaced my use of `(<name> * module_type) option ` by (various flavors of) the type:
```ocaml
type functor_parameter = Unit | Named of <name> * module_type
```
I'm expecting people to cringe at the use of `Unit` there, but that's what the manual uses, and I was really unconvinced by @gasche's "generative argument".
I also didn't add the `Unnamed of module_type` case suggested by @gasche, mostly out of lazyness (because it required minting on `Parsetree` type for structures, and one for signatures).

In this representation too I used `_ option` in the named case, but I'm tempted to backtrack that choice.
Initially I propagated that all the way to `Types`, but it prevented us from strengthening in situations where the parameter is not named, which results in errors such as:
```
# module F(_ : sig end) : sig type t = private int end = struct type t = int end;;
module F : sig  end -> sig type t = private int end
# module G : functor(X : sig end) -> sig type t = F(X).t end = F;;
Error: Signature mismatch:
       Modules do not match:
         sig  end -> sig type t = private int end
       is not included in
         functor (X : sig  end) -> sig type t = F(X).t end
       At position functor (_) -> <here>
       Modules do not match:
         sig type t = private int end
       is not included in
         sig type t = F(X).t end
       At position functor (_) -> <here>
       Type declarations do not match:
         type t = private int
       is not included in
         type t = F(X).t
       A private type would be revealed.
```
So I backtracked that part of the changes (i.e. the one in `Types`) and I'm now minting fresh `_` idents when producing the signature, which feels dirty.
I'm open to suggestions on this point.

I have yet to add tests with unnamed modules.

---
> I doubt it's particularly helpful but I have a branch implementing the `module _ =` part of #6662

It was helpful actually! (to check some of the decisions I made)

**Comment by lpw25** [comment]
> So I backtracked that part of the changes (i.e. the one in Types) and I'm now minting fresh _ idents when producing the signature, which feels dirty.

I think this is not a good approach. Currently we can get nonsense like:
```ocaml
# module G (F : functor (_ : sig end) -> sig type t end) = struct
    module Foo = F
  end;;
    module G :
  functor (F : sig  end -> sig type t end) ->
    sig module Foo : sig  end -> sig type t = F(_).t end end
```

I think the best option is to create a name for a parameter when we strengthen functors with unnamed parameters. That will fix your error example above as well as my example.

**Comment by trefis** [comment]
> I think the best option is to create a name for a parameter when we strengthen functors with unnamed parameters.

I considered that, but didn't know which scope to give to that new identifier.
Thinking more about it, I think giving it the same scope as the path we use to strengthen is acceptable (I'd welcome a second opinion on that point).

Here is what I get on your example after the commits I just pushed:
```ocaml
module G :
  functor (F : sig  end -> sig type t end) ->
    sig
      module Foo : functor (Arg : sig  end) -> sig type t = F(Arg).t end
    end
```

Notes:
- I still haven't added any new tests
- the code in includemod isn't quite right I believe
- tests/shadow_include/shadow_all.ml is broken (probably because of the previous point)

**Comment by Octachron** [comment]
When I was playing #8905, I though about generating the names when typing rather than fiddling with the printing. However, it seemed quite complex to handle cases like: 

```OCaml
 module F(X: sig end -> sig module Arg: sig end type t end) = X;;
```
>  module F :
  functor (X : sig  end -> sig module Arg : sig  end type t end)
    (Arg : sig  end) -> sig module Arg : sig  end type t = X(Arg/2).t end


where it is the names chosen by the type checker that lead to the problematic collision with the names written by the user.

**Comment by trefis** [comment]
I fixed the broken test (which was due to an error in nondep), and improved the includemod code.
I also added some new tests, one of which is currently broken with flambda because I don't know how to provide a dedicated reference file.

That being said, I believe this is ready for review: it's probably easier to review the full diff (especially since some of the commit add erroneous code that is fixed later), but I didn't squash the commits yet.

As expected, there are conflicts in the printer, which should be straightforward to resolve and shouldn't really hinder review. I'll rebase in a couple of days.

**Comment by trefis** [comment]
Rebased.

**Comment by trefis** [comment]
> When I was playing #8905, I though about generating the names when typing rather than fiddling with the printing. However, it seemed quite complex to handle cases like:
> ```
> module F(X: sig end -> sig module Arg: sig end type t end) = X;;
> module F :
>   functor (X : sig end -> sig module Arg : sig end type t end)
>     (Arg : sig end) -> sig module Arg : sig end type t = X(Arg/2).t end
> ```
> where it is the names chosen by the type checker that lead to the problematic collision with the names written by the user.

Indeed. I still find this slightly better than `X(_).t` however.
Also, I guess one question that one might ask is: why does the `/2` only appear on the use of `Arg`, not on the binding?

File: typing/typemod.ml
Comment (trefis) [reply typing/typemod.ml]: Ping @garrigue / @lpw25 .

File: parsing/parsetree.mli
Comment (Drup) [inline parsing/parsetree.mli:732]: Maybe a small comment, in the style of the rest of the file ?

File: parsing/parsetree.mli
Comment (Drup) [inline parsing/parsetree.mli:241]: Same as before, it needs a comment saying that `_` maps to `None`.

**Comment by Octachron** [comment]
@trefis, I agree that `Arg` is much better than  `_` when everything works out. But in case of problems, it is less clear than the name was generated. Maybe a less natural name like `Arg_0` would work better?
Similarly, would it be sensible to have a separated class for generated identifiers (like `Arg` but also `#recmod#`)?
 
Concerning the binding question, since
```OCaml
module type T = sig type t end
module F(X:T)(X:T) = struct type t = X.t end
```
is perfectly valid, it seems reasonable to warn only of uses of the shadowed `X` argument?

File: typing/typemod.ml
Comment (lpw25) [reply typing/typemod.ml]: I agree it shouldn't need re-checking.

**Comment by trefis** [comment]
> Maybe a less natural name like `Arg_0` would work better?

That's fair, we could make it a name that isn't writable by the user (so `#Arg` or `$Arg` or ...).

> Similarly, would it be sensible to have a separated class for generated identifiers (like Arg but also #recmod#)?

Do you mean, adding constructors to `Ident.t`? Who would be looking at that information?

> Concerning the binding question ...

I think whenever you need a `/...` suffix to disambiguate a use, you should also disambiguate the binding. Otherwise it's just confusing.
But sure, if a shadowed id is never used, there is no need to suffix it.

---

Btw, I added *two* changes entry in this PR! Speak up if you think it is excessive.
Also speak up if you can come up with a better name for the PR.

In other news, @Drup kindly accepted an offline request to review this PR.
Thanks Drup!

**Comment by Octachron** [comment]
> Do you mean, adding constructors to Ident.t? Who would be looking at that information?

Indeed. For the anonymous arg, it would be the printer code, which could make the choice to pessimize the name to a syntactically invalid name, or always use the `name+id` format, or try harder (aka #8905). For `#recmod#`, it would be the typechecking code that is trying to catch access to the module types of recursive module inside their definition.

> I think whenever you need a /... suffix to disambiguate a use, you should also disambiguate the binding. Otherwise it's just confusing.

That could be added quite easily. In fact, I think that there is small subset of #8905 that would do exactly that.

**Comment by lpw25** [comment]
(I think I already removed the `#recmod#` stuff)

**Comment by Octachron** [comment]
@lpw25 , ah indeed. This is was bad example, sorry.

File: typing/typemod.ml
Comment (trefis) [reply typing/typemod.ml]: Thanks, I pushed a commit removing that (and another similar) call.

**Comment by trefis** [comment]
@Octachron I'm not particularly convinced by your suggestion, but more importantly: I'm unwilling to give it a try here (because I think this PR is already big enough, and because I don't have many spare cycles at the moment).

@Drup: I know you had started reviewing this a while back, how is it looking? :)

**Review by Drup — APPROVED** [review]
I was somehow slightly surprised: I don't really have any comments. The changes are mostly boilerplate/reorgnazation, and I agree with the general design and use-cases.

The only change I would ask for are to make sure `parsetree.mli` file is still equally documented with examples for the few changes.

**Comment by trefis** [comment]
I added the comments @Drup suggested in `parsetree.mli`, and rebased.
I'll merge once the CI comes back green.

Thanks @Drup!

**Review by Drup — APPROVED** [review]

**Comment by Octachron** [comment]
@trefis , a point that I had missed is that this PR changes the name of extension constructors defined inside a generative functor:

```OCaml
F() = struct exception E end
```
is now translated to
```lisp
(let (E/87 = (makeblock 248 "E" (caml_fresh_oo_id 0)))
```
rather than
```
(let (E/88 = (makeblock 248 "F(*).E" (caml_fresh_oo_id 0)))
```
previously.

Should we not keep the functor path in the extension constructor name (i.e. `F().E` ) ?

**Comment by trefis** [comment]
Are you referring to the following point of my initial message:
> the code in translmod that creates path like F(*).Foo (probably) wouldn't have been written in the first place?

I'm pretty sure I was convinced it was a good idea back then.
Right now :man_shrugging: 

More precisely: this doesn't really matter for anything, and I believe only affects the bytecode and flambda backends.
In those cases: it's possible that some "Uncaught exception: ..." error messages which were previously printing `F(*).E` will now print "E"; but I haven't checked (if they did print that before).
(Also: people compiling their code with flambda and an hypothetical dwarf/gdb support, and then debugging it in gdb might or might not see "E" instead of `F(*).E`?)

Again, at the time I thought this was a positive change. Now I still think it's a reasonable change, but wouldn't object if someone proposed that "F().E" be printed instead, as long as their is a clean implementation.

Does that sound reasonable to you? Am I missing something?

**Comment by Octachron** [comment]
Indeed, I was referring to this point, and indeed the error message for uncaught exception in
```OCaml
module F() = struct exception E end
let _ =
  let open F() in
  raise E
```
is right now:

> Fatal error: exception E

rather than

> Fatal error: exception Filename.F(*).E

At least this is the case on ocamlopt (without flambda) and ocamlc.
So it seems that they are some information lost on this narrow corner case.
It doesn't seem that important but I may have a look ... later.

**Comment by gasche** [comment]
For some context, @Octachron and myself looked at this together while he was working on #9021: that PR touches code in genprintval that parses the runtime name of extensions as a constructor, but in fact using a slightly different grammar (in OCaml programs themselves you cannot write `F(X).Foo` to denote a constructor, we only accept "simple" module paths).

The change in the present PR (removing the functor application for generative functors) does not make this behavior go away (it remains for applicative functors), and in fact this behavior is useful to name uncaught exceptions (we might even get more type information than we have today if we tried harder), and I agree with the two of you that it would be better to have `F().E` if reasonably possible.

**Comment by yallop** [comment]
It'd be good to update the manual to document that `_` can now appear in place of a module name in binding contexts such as `module _ = M`

**Pull request merged**
